keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1514 @@
+need for reinvocation. Also, in-process mutation is sufficiently fast,
+especially when compared with webhooks, that it is reasonable to re-run
+mutations to do things like validate that multiple muation policy is still
+applied after all other mutating admission operations have been applied.
+
+### Goals
+
+- Provide an alternative to mutating webhooks for the vast majority of mutating
+  admission use cases.
+- Provide the in-tree extensions needed to build policy frameworks for
+  Kubernetes, again without requiring webhooks for the vast majority of use
+  cases.
+- Provide an out-of-tree implementation of this enhancement (using a webhook)
+  that is supported by the Kubernetes org to provide this enhancement
+  functionality to Kubernetes versions where this enhancement is not available.
+- Provide core functionality as a library so that use cases like GitOps,
+  CI/CD pipelines, and auditing can run the same CEL validation checks
+  that the API server does.
+
+### Non-Goals
+
+- Build a comprehensive in-tree policy framework. We believe the ecosystem is
+  best equipped to explore and develop policy frameworks.
+- Full feature parity with mutating admission webhooks. For example, this
+  enhancement is not expected to ever support making requests to external
+  systems.
+- Replace the admission controllers compiled into the API server. 
+- Static or on-initialization specification of admission config. This is a
+  needed feature but should be solved in a general way and not in this KEP
+  (xref: https://github.com/kubernetes/enhancements/issues/1872).
+
+## Proposal
+
+### Summary
+
+Before getting into all the individual fields and capabilities, let's look at the general "shape" of the API.
+Very similar to what we have in ValidatingAdmissionPolicy, this API separates policy definition from policy configuration by splitting responsibilities across resources. The resources involved are:
+- Policy definitions (MutatingAdmissionPolicy)
+- Policy bindings (MutatingAdmissionPolicyBinding)
+- Policy param resources (custom resources or config maps)
+
+The idea is to leverage the CEL power of the object construction and allow users to define how they want to mutate the admission request through CEL expression.
+This proposal aims to allow mutations to be expressed using the "apply configuration" introduced by Server Side Apply. 
+And users would be able to define only the fields they care about inside MutatingAdmissionPolicy, the object will be constructed using CEL which would be similar to a Server Side Apply configuration patch and then be merged into the request object using the structural merge strategy. 
+See sigs.k8s.io/structured-merge-diff for more details.
+
+Note: See the alternative consideration section for the alternatives.
+
+Pros:
+- Consistent with Server Side Apply so that we will continue investing SSA as the best way to do patch updates to resources;
+- Does not require the users to learn a new syntax;
```

Comment:
Is the CEL object instantiation syntax used already in validating admission policy? /where is it already used that it would be familiar 

To me this is new, and my first impression is that it looks rather clunky. Are we expecting end users who need to enforce policies for their clusters, to be familiar with this syntax already?

Reply:
The object instantiation is not used in validation since we normally ask for a true/false return for validation. But VAP should hopefully help people familiar with CEL syntax which reduce the high learning curve. Even from scratch, CEL is designed to be simply with no surprising syntax. :)

Reply:
My initial thought of the new syntax was that having to specify the types structure, almost like it's go code was rather verbose

Compare
```
Object{
  spec: Object.spec{
    containers: object.spec.containers.map(c,
      Object.spec.containers.item{
          name: c.name,
          imagePullPolicy: "Always"
      }
    )
  }
}
```
To something without the type declarations, plain YAML for example, I wondered if something simpler could be created to achieve the same result, but I expect the implementation of that would be equivalent to a new language in itself

Reply:
Yes we have considered that direction but it will make the conditional statement hare to write(like I only wanna to modify containers whose name is not begin with `demo` or I only wanna create a patch when volumes are missing from volumeMount). And I agree it will be equivalent to a new language in itself if we add those support -_-